Updated to Debian 13-slim base, security improvements, removed forced creds.

kevinruffus opened the pull request:
1. Added example env file to allow for assignable info and creds
2. Updated docker-compose files to reflect env file addition and use updated version options
3. Now copies example json and ini files for non mlapi version to correct location and enables persistence
4. Defaults to using docker managed persistent volumes
5. Uses current MariaDB image and adds basic healthcheck function
6. Updates Zoneminder to release 1.36.35
7. Updates base image to Debian 13-slim
8. Bumps PHP ver to 8.4
9. Added a few php and apache2 security settings
10. Added Nano to the image because not being able to edit files while attached to the container was absolutely obnoxious.

After or before building you'd need to update the image listed in the docker-compose files to pull from your repo to match the release that's created.

kevinruffus committed:
Added example .env file, renamed es_rules and ini files in root and added dockerfile entries to copy them into correct locations, removed static ENV variables from Dockerfile, pushed Debian base to 12.12, updated compose files to use .env file and to use updated mariadb container with a healthcheck. External exposed ports can now be assigned via the .env file as well. Users should not need to alter the compose files unless they use specific networking configurations.

kevinruffus committed:
Added persistence to config files in /etc/zm via dockerfile, enrtypoint, and compose file changes.

kevinruffus committed:
Fixed spacing in if statements in entrypoint

kevinruffus committed:
Fixed overcomplication of persistent config files

kevinruffus committed:
Updated to Debian 13 slim to reduce size and solve CVEs

kevinruffus committed:
Updated entrypoint command to reflect update to php 8.4 with new base image.

kevinruffus committed:
Testing security conf updates and memory of sed commands

kevinruffus committed:
attempting to resolve syntax error in security.conf sed config commands

kevinruffus committed:
next attempt at sed commands inserting after match

kevinruffus committed:
undo last commit

kevinruffus committed:
syntax fix, hopefully last

kevinruffus committed:
Added a few php hardening entries

kevinruffus committed:
Removed Same-origin as it can cause issues.

kevinruffus committed:
Testing fix for Delete not working

kevinruffus commented:
Hold on putting this through. Checking something that may be an unintended side effect of the updates.

kevinruffus committed:
Fully removed security additions for apache and php as they broke certain functions, including manual deletion through the events page.

kevinruffus commented:
Removed problematic security settings from php.ini and security.conf
Issue now resolved.

kevinruffus committed:
Added security.conf settings back after testing to determine which were NOT causing the delete button issue.

jantman commented:
@kevinruffus thanks so much for this. I gave a try at using Copilot's code review... full disclosure I've never tried that for a PR from someone else before. The one thing I'm confused about is the removal of the EXAMPLE files... I'm not sure I follow that, and I agree with copilot that it seems like it should cause the Docker build to fail?

kevinruffus commented:
@jantman They actually weren't removed, but instead just renamed and dumped in place, as they would have no real effect unless the option was enabled in the settings through the UI. That being said, I think I have a better idea. A simple ENV variable to "enable" ES, which then runs checks for the files with EXAMPLE in the name, and renames them if they exist. If they don't, it moves on, thus allowing persistence. Simple enough to do. I'll get some of the other things cleaned up as well.

kevinruffus commented:
@jantman Oh! The mlapi ini files. That was a mistake. Thank you for pointing it out.

kevinruffus committed:
Removed white spaces, added ENV variable to enable ES and entryfile logic for it, fixed mlapi filename references.

Kasslim commented:
I did a quick read-through of the diffs and found nothing crazy in the process (keep in mind I'm new at containerization however).

I do want to mention that the following files have diffs on every single line:
- es_rules.EXAMPLE.json → es_rules_EXAMPLE.json
- zmeventnotification.EXAMPLE.ini → zmeventnotification_EXAMPLE.ini

I checked and it's because Windows line endings were added (\r\n as opposed to \n). I don't know which line endings are normal for this project, but would generally expect only \n. Git can be configured to fix these automatically when making commits.
I also spotted some erroneous whitespace at the end of several lines. These could probably be removed automatically by your code-editor depending on which one you use (e.g. VSCode).

Can you remediate these (minor) issues? It makes viewing diffs easier and helps keep everything consistent.

Otherwise nice looking changes! I figured I'd review as part of my learning-journey.

kevinruffus committed:
Fixed end of line issues... again.

kevinruffus commented:
@Kasslim 
Thank you for pointing out the end of line issue.
VS Code on Windows defaults to crlf, and it's the second time I've had to fix that now. It must have reset them when I renamed the files the last time. I assume it's actually removing the original files then recreating the "new" renamed file using whatever settings are in Code vs what was in the original file.
This time I set it to all profiles and went through and corrected the affected files.

As to the file names, I went back and forth on what to do with them, ultimately deciding on "enabling" them via an env variable and a bit of logic in the entryfile. I had originally moved them as part of the default setup but realized there wasn't any reason to do so if event notification wasn't going to be used. It looks like I just typed a _ instead of a . when I renamed them the last time.

It's been a while since I've written anything but PowerShell scripts, so I'm a bit rusty myself, and this is the first time I've actually built a container. I've helped with the bash scripting in the entryfile and dockerfile before, but this is my first actual complete build.

Kasslim commented:
Happy to help!

I'm excited to try out this repo when I have a little bit of time, perhaps this weekend. I hope to set it up for long term use but the instructions seemed pretty high level (for me), so perhaps I'll write a little guide with commands and explanation as I go. If you already happen to have notes that might boost me along those would be very welcome, but if not no worries. (I also don't know how you'd even share them, since it's off topic for this PR.)

Either way, thanks jantman for setting up the repo and you for working to improve it!

jantman commented on Dockerfile:
```diff
@@ -71,6 +65,14 @@ RUN install -m 0644 -o root -g root /tmp/zm-site.conf /etc/apache2/sites-availab
 VOLUME /var/cache/zoneminder
 VOLUME /var/log/zm
+VOLUME /var/log/apache2
+VOLUME /etc/zm
+
+# Copy default files for ZMES
+COPY es_rules_EXAMPLE.json /etc/zm/
+COPY secrets_EXAMPLE.ini /etc/zm/
+COPY zmeventnotification_EXAMPLE.ini /etc/zm/
+COPY objectconfig_EXAMPLE.ini /etc/zm/
```
Maybe this is a personal preference, but I do not want a volume for the config - the intent is for it to be mounted in from the host.

jantman commented on Dockerfile:
```diff
@@ -37,6 +30,7 @@ RUN apt update \
          python3-opencv \
          s6 \
          wget \
+         nano \
```
Nano has no place inside a container any more than it has a place inside the package of another piece of software.

jantman commented on entrypoint.sh:
```diff
@@ -28,7 +28,32 @@ chmod -R 770 /etc/zm /var/log/zm
+sed -i "s/ServerSignature On/ServerSignature Off/" /etc/apache2/conf-available/security.conf
+sed -i 's/#Header set X-Content-Type-Options: "nosniff"/Header set X-Content-Type-Options: "nosniff"/' /etc/apache2/conf-available/security.conf
+
+# Enable event notification if requested
+if [ -e /etc/zm/es_rules_EXAMPLE.json ] && [ "$ENABLE_ES" == "true" ]; then
+    mv /etc/zm/es_rules_EXAMPLE.json /etc/zm/es_rules.json
+fi
+
+if [ -e /etc/zm/objectconfig_EXAMPLE.ini ] && [ "$ENABLE_ES" == "true" ]; then
+    mv /etc/zm/objectconfig_EXAMPLE.ini /etc/zm/objectconfig.ini
+fi
+
+if [ -e /etc/zm/secrets_EXAMPLE.ini ] && [ "$ENABLE_ES" == "true" ]; then
+    mv /etc/zm/secrets_EXAMPLE.ini /etc/zm/secrets.ini
+fi
+
+if [ -e /etc/zm/zmeventnotification_EXAMPLE.ini ] && [ "$ENABLE_ES" == "true" ]; then
+    mv /etc/zm/zmeventnotification_EXAMPLE.ini /etc/zm/zmeventnotification.ini
+fi
```
This doesn't make sense - these blocks overwrite the user's config files with the useless example configs. The example config files are a reference for humans - you copy the example config file to the real config file, and then edit as needed.

jantman reviewed (commented):
Just getting back to look at this, sorry for the long delay.

I've made a few comments, but the main one is around the use of volumes vs bind mounts. I set this up to use bind mounts and would like to keep it that way - volumes are a pain to backup and restore, a pain to get access to the files (maybe why nano was added?), and lots of people don't understand them and end up losing either their data or disk space or both. Bind mounts are easy and work the way people expect filesystems to work, and can be backed up and restored like any other files.

jantman commented:
I appreciate all of the contributions here but ZM 1.38.0 is out and I'm in the process of updating to that version in #10 ; that will be the new version of this repo, and will be based on Trixie.
